lib/src/widgets/typing_indicator.dart

```diff
@@ -0,0 +1,487 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_chat_types/flutter_chat_types.dart' as types;
+import '../../flutter_chat_ui.dart';
+import '../util.dart';
+import 'state/inherited_chat_theme.dart';
+
+class TypingIndicator extends StatefulWidget {
+  const TypingIndicator({
+    super.key,
+    required this.bubbleAlignment,
+    this.options = const TypingIndicatorOptions(),
+    required this.showIndicator,
+  });
+
+  /// See [Message.bubbleRtlAlignment].
+  final BubbleRtlAlignment bubbleAlignment;
+
+  /// See [TypingIndicatorOptions].
+  final TypingIndicatorOptions options;
+
+  /// Used to hide indicator when the [options.typingUsers] is empty.
+  final bool showIndicator;
+
+  @override
+  State<TypingIndicator> createState() => _TypingIndicatorState();
+}
+
+class _TypingIndicatorState extends State<TypingIndicator>
+    with TickerProviderStateMixin {
+  late double stackingWidth;
+  late AnimationController _appearanceController;
+  late AnimationController _animatedCirclesController;
+  late Animation<double> _indicatorSpaceAnimation;
+  late Animation<Offset> _firstCircleOffsetAnimation;
+  late Animation<Offset> _secondCircleOffsetAnimation;
+  late Animation<Offset> _thirdCircleOffsetAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    _appearanceController = AnimationController(
+      vsync: this,
+      duration: const Duration(milliseconds: 500),
+    );
+
+    _indicatorSpaceAnimation = CurvedAnimation(
+      parent: _appearanceController,
+      curve: const Interval(0.0, 1.0, curve: Curves.easeOut),
+      reverseCurve: const Interval(0.0, 1.0, curve: Curves.easeOut),
+    ).drive(Tween<double>(
+      begin: 0.0,
+      end: 60.0,
+    ));
+
+    _animatedCirclesController = AnimationController(
+      vsync: this,
+      lowerBound: 0.0,
+      upperBound: 1.0,
+      duration: widget.options.animationSpeed,
+    )..repeat();
+
+    _firstCircleOffsetAnimation = _circleOffset(
+      Offset.zero,
+      const Offset(0.0, -0.9),
+      const Interval(0.0, 1.0, curve: Curves.linear),
+    );
+    _secondCircleOffsetAnimation = _circleOffset(
+      Offset.zero,
+      const Offset(0.0, -0.8),
```

Comment:
is this lower different (0.8 vs 0.9) on purpose?

Reply:
Yes, these offsets represent transitions of circles on y-axis.